drv/stm32h7-spi-rot-server/src/main.rs

```diff
@@ -0,0 +1,274 @@
+    // The larger of the two buffers.
+    let message = &mut [0u8; drv_spi_msg::SPI_RSP_BUF_SIZE];
+    let mut server = ServerImpl {
+        sys,
+        spi,
+        message: *message,
+    };
+
+    loop {
+        idol_runtime::dispatch(&mut buffer, &mut server);
+    }
+}
+
+fn do_send_recv(server: &mut ServerImpl) -> Result<usize, MsgError> {
+    ringbuf_entry!(Trace::Data(
+        server.message[0], // Protocol == 1
+        server.message[1], // Payload size LSB
+        server.message[2], // Payload size MSB
+        server.message[3]  // MsgType
+    ));
+    let msg = Msg::parse(&mut server.message[..]).unwrap_lite();
+    if !msg.is_supported_version() {
+        return Err(MsgError::BadMessageType);
+    }
+    match msg.msgtype() {
+        MsgType::EchoReq | MsgType::SprocketsReq | MsgType::StatusReq => (),
+        _ => {
+            ringbuf_entry!(Trace::Line);
+            return Err(MsgError::BadMessageType);
+        }
+    }
+    let xmit_len = SPI_HEADER_SIZE + msg.payload_len();
+    ringbuf_entry!(Trace::Line);
+
+    let port = server.sys.gpio_read_input(ROT_IRQ.port).unwrap_lite();
+    let rot_irq = port & ROT_IRQ.pin_mask;
+    // let rot_irq = server.sys.gpio_read(ROT_IRQ).unwrap_lite();
+    if 0 == rot_irq {
+        // We are surprised that ROT_IRQ is asserted
+        // It is intended to be ok to ignore RoT responses.
+        //
+        // The RoT must be able to observe SP resets.
+        // During the normal start-up seqeunce, the RoT is controlling the
+        // SP's boot up sequence. However, the SP can reset itself and
+        // individual Hubris tasks may fail and be restarted.
+        // TODO: Should this task send an explicit message telling the RoT
+        // that the SP has fully initialize? The first Sprockets message may
+        // serve as that indication.
+        //
+        // If the RoT ever does asynchronous transmission of messages
+        // then we'll see ROT_IRQ asserted without having first sent a message.
+        // If SP and RoT are out of sync, e.g. this task restarts and an old
+        // response is still in the RoT's transmit FIFO, then we can also see
+        // ROT_IRQ asserted when not expected.
+
+        ringbuf_entry!(Trace::GpioPort(port, ROT_IRQ.pin_mask));
+        // Async and unhandled responses from RoT are not expected.
+        panic!();
+    }
+    if let Err(spi_error) = server.spi.write(&server.message[0..xmit_len]) {
+        ringbuf_entry!(Trace::SpiError(spi_error));
+        return Err(MsgError::SpiServerError);
+    }
+
+    // We sleep and poll for what should be long enough for the RoT
+    // to queue a response.
+    // TODO: For better performance and power efficiency,
+    // take an interrupt on ROT_IRQ falling edge with timeout.
+    const LIMIT_START: u64 = 1_000_000; // TODO: Something more reasonable.
+    const NAP: u64 = 1;
+    let mut limit = LIMIT_START; // 1.04ms Time limit is somewhat arbitrary.
```

Comment:
`sleep_for` is in terms of kernel ticks, not clock ticks. Each kernel tick is 1 ms, so this will actually wait for 16 minutes!

Reply:
Yeah. That wasn't right. Hence the TODO.